### ISSUE TITLE
Change history does not update for a downloadable resource

### ISSUE THREAD

**Issue by genkimaps** [genkimaps/gis-hub#41]
This enhancement will require moving the change history field to the resource metadata schema, unless there is a way to trigger an update of the _dataset-level_ change history field when a resource's data is updated. 

Ideally this should be tied to the action of uploading a new data file.  

Related to:  https://github.com/genkimaps/gis-hub/issues/88

Edit: changed the title of this issue to reflect what I found out.  This is a bug, see last comment.

**Comment by genkimaps** [comment]
Might work if we simply move the change history field into the resource section of the custom metadata schema, and make it a required field. Think about how to display past change history (while not making past entries editable).  When editing schema, made a new field called resource_change_history and leave the old change history at the dataset level in place for now, or existing datasets will throw an error with the new schema.

**Comment by genkimaps** [comment]
~~Do we also need to make past entries non-editable? No, we should keep them editable, in case their details need to be changed later.~~  Disagree, I think this will clog up the dataset edit page and cause "form fatigue" among users. We want them to focus on a single change history field, simply providing the rationale for the current change, whatever it may be. 

First, we should make this into a table display, similar to attributes in the resource metadata.  Currently we have _Description of Change_ and _Date of Change_ stacked vertically in the dataset display, which is confusing.

**Comment by genkimaps** [comment]
When I changed the change history field to a table display, I noticed that the column spacing is not ideal.  This is fixed with this change:  https://stackoverflow.com/a/52463416.  Added the `table-fit` class to the `table.html` definition in ckanext-scheming, and updated `dfo.css` in ckanext-dfo. 

Before:
![image](https://user-images.githubusercontent.com/15989988/74501210-ba3df780-4f2c-11ea-84c4-7f5b8a67a784.png)

After:
![image](https://user-images.githubusercontent.com/15989988/74501377-4e0fc380-4f2d-11ea-9e80-e9b9d766751b.png)

**Comment by genkimaps** [comment]
Best option:
Keep the existing change_history field at the dataset level, but make it read-only, in other words **remove the form snippet** in the schema file.  Add new fields:

1. change_description_dataset, a required field(?) at the dataset level, user must describe what was changed in metadata
2. change_description_resource, a required field at the resource level, user must describe what was changed in metadata **and/or** in the uploaded file

When an updated dataset or resource is submitted, this triggers the `object_updated_or_created()` function in the plugins module in ckanext-dfo.  Take whatever change history value was submitted in change_description_* (either one will be present, depending if it's a resource or a dataset) and append it to the change_history field at the dataset level with the current date. 

Will also need to update hub-geo-api to ensure that when a dataset/resource is updated internally, this change history is _not_ updated--I think we only want the change history to reflect user-initiated changes, not things that are auto-populated behind the scenes. Since the change_description_* fields are required, hub-geo-api should insert a dummy value like 'internal update from hub-geo-api' which is ignored by ckanext-dfo.

**Comment by genkimaps** [comment]
Note: do **not** modify the existing change_history field at the dataset level. Keep it as-is, in particular, keep the existing form snippet. Remove the change_history_dataset field, it is redundant and confusing.  Users can update change history for the **dataset** using the existing form fields, whereas changes for **resources** will be appended to the dataset-level change history automatically.

**Comment by genkimaps** [comment]
This feature is deployed, though it's not quite ideal. Making the change description field for resources required does not work, for some reason, even though it uses the same validator as elsewhere in the system.  In any case, change descriptions are appended to the dataset-level change history field.

**Comment by cole-fields** [comment]
Small css formatting issue with the table:
![image](https://user-images.githubusercontent.com/52669482/74657639-aa0d5e80-515e-11ea-9335-aab1c0a33298.png)

In https://www.gis-hub.ca/dataset/substrate100m-data which had several existing updates.
So, in this case, the change updates were kept at the dataset level because they existed before the for change, but if I were to update it now, the changes would be appended to the dataset rather than the resourse?

**Comment by genkimaps** [comment]
@cole-fields 
I have a separate issue for the table layout: https://github.com/genkimaps/gis-hub/issues/68

> if I were to update it now, the changes would be appended to the dataset rather than the resourse?

Previously, only changes at the dataset level were recorded (using the change history field).  Now, a change comment entered when updating a resource will also be recorded in the _dataset's_ change history.   This makes sense to me, given that the vast majority of updates to resources will be cases where the uploaded zip file is being replaced with a new version of the data.  In this case, it seems reasonable to attach the change comment to the _dataset_, since there is only supposed to be one upload resource per dataset (the others are auto-generated).  However, I can understand how users might find it confusing.  What do you think?

**Comment by cole-fields** [comment]
Yes, your explanation makes sense to me and I think you're correct in assuming most of the updates will be related to the dataset. Should be useful.

> since there is only supposed to be one upload resource per dataset 
By that, you mean the single unploaded zip file, correct? But inside that zip file, there may be several datasets.

**Comment by cole-fields** [comment]
noticed that there was no change history documented in the metadata when i updated this dataset: https://www.gis-hub.ca/dataset/env-layers-nsbssb either at the dataset or resource level. could this be because it is set to private?

**Comment by genkimaps** [comment]
@cole-fields 
Did you fill in the Change History field manually? It does not get filled in automatically.  When editing, there are change history fields at both resource and dataset level, but the history is shown only at dataset level.  I just added entries for both resource and dataset:  https://www.gis-hub.ca/dataset/env-layers-nsbssb

**Comment by cole-fields** [comment]
Seems to work now. When I tried before, it was while also updating the dataset and it didn't capture my entry for the Change History.

**Comment by genkimaps** [comment]
@cole-fields Is this working for you now? Can't remember if we discussed this

> When I tried before, it was while also updating the dataset and it didn't capture my entry

How did you update the dataset and resource at the same time? Those are two separate pages. 

> Does it check to see if inside the zip there are new files and only appends the change history text if that is true?

No, it relies on the user to enter text in the "Description of Change" field.

**Comment by cole-fields** [comment]
@genkimaps yes, this is working great now. I must have been doing something incorrectly.

**Comment by genkimaps** [comment]
Issue resolved, can revisit if users find it confusing.

**Comment by cole-fields** [comment]
@genkimaps myself and others have found this confusing or not functioning as expected. my testing:
 - updates at the dataset level (adding a Change History entry) works
 - updates at the resource level are not working (neither with re-uploading a new zip file nor without)

I like the idea of making it mandatory for users to supply a change history entry, but doesn;t seem to be working. Perhaps we remove the change history at the resource page?

**Comment by genkimaps** [comment]
@cole-fields 
This is next on my list, will investigate why the resource level changes are not saved.  I added the required field at the resource level because our expectation was that users are lazy and won't enter anything unless required.  But since you and Kayleigh are kind of acting as gatekeepers, maybe it won't be a problem remove it at the resource level, and add a reminder for users to update Change History for the dataset.

**Comment by genkimaps** [comment]
@cole-fields  I don't know the cause of this, but the behavior I have seen is that for resource metadata fields defined in our custom schema (i.e. fields that don't exist in resources in a default CKAN setup), changes do not stick for the first resource, which is the download placeholder.  

Digging into the code, the before_update function in CKAN (https://docs.ckan.org/en/2.9/extensions/plugin-interfaces.html#ckan.plugins.interfaces.IResourceController.before_update) has the correct metadata, but after it is saved in the DB, it reverts to the default value, or it gets replaced by the value of the same field in a subsequent resource.  

For example, notice that the first resource in this dataset: https://www.gis-hub.ca/api/3/action/package_show?id=test20201009
Has fields "data_format": "GeoTiff", "filepath": "tiffimage.tif" and so on, which does not make sense, since this is only a download placeholder resource.  All of the geospatial resource metadata fields should be blank for that resource.  These metadata fields are being overwritten somewhere in the guts of CKAN, but only for the first resource in the list of resources.  If we can fix this core bug, it will likely also resolve https://github.com/genkimaps/gis-hub/issues/88, which suffers from the same problem. 

It does not matter if I update the resource's metadata through the API, or through the web interface.  Both have the same problem.  This might be a side effect of the ckanext-scheming extension, but I'm not sure.

**Comment by cole-fields** [comment]
@genkimaps - interesting. That is similar to what I discovered when looking at restricting resources to specific users. I could alter one resource in the dataset and the top level (download placeholder), but those changes wouldn't stick. That meant unauthorized users could still download the data which didn't make sense. It only worked as expected once I changed ALL the resources that were part of a dataset to restricted resources. That was when you (and others) got a flurry of emails from the gis hub).

Is it possible that one would need to patch each of the dataset's resources (including the download placeholder) with the update to the metadata for it to stick?

This might end up being a tricky one to fix? But as you mention, might resolve some other functional issues. Thanks for taking a look at this.

**Comment by genkimaps** [comment]
@cole-fields 
An update on this: did some more digging in the code, but I still can't figure out the cause of this. Updating all of the dataset's resources may work, but it's kind of an ugly hack and best avoided if possible.  I have a couple more ideas where this bug may be coming from, which I'll test once I'm back at my Linux workstation.

**Comment by cole-fields** [comment]
@genkimaps sounds good, thanks for the update. I agree - might not be the best solution to update all the resources if we don't have to.

**Comment by genkimaps** [comment]
@cole-fields I think I cracked the root cause of this problem, see this change: 
https://github.com/genkimaps/ckanext-scheming/commit/f03fc1c797b7b453db92eef56e34c0d3cc477c9a
It's live now on the server.  Updating the disclaimer or disclaimer URL seems to work now.  The resource change history field is still disabled, so I can't test that at the moment.

**Comment by cole-fields** [comment]
@genkimaps the download disclaimer and URL seem to work (i closed that issue). I tested bringing back the resource level description of change and didn't have success with it - see branch res_changehist

**Comment by genkimaps** [comment]
@cole-fields FYI I am working on restoring the change history behavior, tried to do this in a slightly different way this time.  It's in branch restore_changehist.  There are some bugs, so I have the ckanext-dfo extension running on master branch for now.

**Comment by cole-fields** [comment]
@genkimaps excellent, thanks for the heads up. There are always some unforeseen bugs! Might be something to discuss on our next call - I'll set up a meeting for a couple weeks from now.

**Comment by genkimaps** [comment]
@cole-fields 
The previous functionality has been restored, please note the following:
- ckanext-dfo is running on branch restore_changehist at the moment
- resource change history for the upload resource is now saved correctly, and it is added to the change history section of the dataset page.  When added to the change history section, the change description also has a label showing which resource it came from.  
- I tried to modify the resource_form template to have a different subset of fields for upload resources and layer/file resources.  In particular, I tried to hide the entire 'Upload Data' section on non-upload resources.  Because of the way the template is structured, and some hard-coded requirements in core CKAN, this did not work.  The best I could do was hide the geospatial fields on the upload resource (which is how it has always been). 
- I discovered that validators in the schema do not work for resource fields, nor for composite fields.  This is annoying, but we have lived with it for a while now, and it doesn't cause any critical problems.  It's a bug in ckanext-scheming, but I haven't tried to fix it yet.  Issue:  https://github.com/genkimaps/ckanext-scheming/issues/2
- For this reason, I removed all validators on the Change Description (Resource) field, and added more details in the help text for that field.   I hope that users will find it less confusing.  I also updated the docs page.

**Comment by cole-fields** [comment]
@genkimaps this looks good, thanks for making this a priority. I like the addition of the resource name added to the change history entry on the dataset metadata page. The only bug I came across in testing was the following:
 - a dataset exists with resources uploaded
 - this dataset has _no existing change history entries_
 - user goes to resource page and uploads a new version of zip file, fills an entry for description of change field
 - this entry is _not added_ to the the change history metadata
But...
 - if there is already at least one entry in the change history, then it works as expected. 
 - i thought of suggesting adding a default value in the desc of change field (such as 'Initial Upload'), but the problem with that is that text shows up once for each resource in the data package if it's the first upload
 - so every dataset needs to have a previous entry in that field from what i can tell

**Comment by genkimaps** [comment]
@cole-fields  This bug was probably caused by incorrectly checking for a null/empty list change history field.  Some older datasets might have been created before we added that field to the schema.  The default value should be `"[{\"change_date\": \"\", \"change_description\": \"\"}]"` (a string representation of a dict with one blank entry), but even if it doesn't exist, we can create it.  

I removed the code that incorrectly rejects a blank change history field.  Please try again.  In the future, when you find a bug like this, please include the dataset ID with the description of the problem.  I wasn't able to reproduce this since we were using different test datasets.  

If you want to see what's happening in the logs for the ckanext-dfo extension, use this script:
`/home/dfo/hub-geo-api/scripts/dfoext_logs.sh`

**Comment by genkimaps** [comment]
Also removed all instances of 'internal update from hub-geo-api' in the change_description_resource fields across all resources on the GIS Hub (change_description_resource is now empty string).  See related issue: https://github.com/genkimaps/hub-geo-api/issues/24

Change was made in commit: https://github.com/genkimaps/hub-geo-api/commit/d44a255e6f853398074dbc34a5675f841ae115b9

**Comment by cole-fields** [comment]
@genkimaps thanks, I just tested on this dataset https://www.gis-hub.ca/dataset/e2fd899b-0db9-4930-a24a-22ea88d6015b? which had  resources already updated, but had no change history entry. I did not encounter the bug I did last time, and so your fix seems to be working. 

> removed all instances of 'internal update from hub-geo-api' in the change_description_resource fields across all resources on the GIS Hub

Excellent, that's a nice update as the text has shown up on many datasets but doesn't really tell the users anything. 

Nice work! I guess we can close this issue and merge the branch with master? I'll leave that for you in case you're still working on this branch.
